extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java

```diff
@@ -144,6 +155,113 @@ public Long terminatePartial(RamAccounting ramAccounting, HllState state) {
+                        // => mix64(doubleToLongBits(NumericUtils.sortableLongToDouble(values.nextValue())))
+                        // => mix64(doubleToLongBits(Double.longBitsToDouble(sortableDoubleBits(values.nextValue()))))
+                        // => mix64(sortableDoubleBits(values.nextValue()))
+
+                        var hash = BitMixer.mix64(NumericUtils.sortableDoubleBits(values.nextValue()));
+                        state.addHash(hash);
+                    }
+                );
+            case IpType.ID:
+            case StringType.ID:
+                var precision = optionalParams.size() == 1 ? (Integer) optionalParams.get(0).value() : HyperLogLogPlusPlus.DEFAULT_PRECISION;
+                return new HllAggregator(fieldTypes.get(0).name(), dataType, precision);
+            default:
+                return null;
+        }
+    }
+
+    private static class HllAggregator implements DocValueAggregator<HllState> {
+
+        private final String columnName;
+        private final DataType<?> dataType;
+        private final Integer precision;
+        private final MurmurHash3.Hash128 hash128 = new MurmurHash3.Hash128();
+        private SortedBinaryDocValues values;
+
+        public HllAggregator(String columnName, DataType<?> dataType, Integer precision) {
+            this.columnName = columnName;
+            this.dataType = dataType;
+            this.precision = precision;
+        }
+
+        @Override
+        public HllState initialState(RamAccounting ramAccounting, MemoryManager memoryManager, Version minNodeVersion) {
+            var state = new HllState(dataType, minNodeVersion.onOrAfter(Version.V_4_1_0));
+            return initIfNeeded(state, memoryManager, precision);
+        }
+
+        @Override
+        public void loadDocValues(LeafReader reader) throws IOException {
+            values = FieldData.toString(DocValues.getSortedSet(reader, columnName));
+        }
+
+        @Override
+        public void apply(RamAccounting ramAccounting, int doc, HllState state) throws IOException {
+            if (values.advanceExact(doc) && values.docValueCount() == 1) {
+                BytesRef ref = values.nextValue();
+                // Using ref.bytes for getting utf-8 bytearray is wrong,
+                // ref.bytes and DataTypes.STRING.implicitCast(ref).getBytes(StandardCharsets.UTF_8)
+                // produce different arrays with same beginnings but ref.bytes has some "unused" bytes at the end.
+                // BytesRef has a property 'length' providing length of used bytes.
+                // Offset has to be taken to account as well.
+
+                byte[] bytes = new byte[ref.length];
+                System.arraycopy(ref.bytes, ref.offset, bytes, 0, ref.length);
+
+                var hash = state.isAllOn4_1() ?
+                    MurmurHash3.hash64(bytes, bytes.length)
+                    : MurmurHash3.hash128(bytes, 0, bytes.length, 0, hash128).h1;
```

Comment:
How difficult would it be to adapt the hash functions to support a offset parameter to avoid the array copy?

Reply:
Updated hash64 method. Hash128 already works with offset

Reply:
Re-run becnhmark, updated PR description with the new result -  it got better, thanks.

Still don't understand why second conpary_py makes 2 rounds and second round is faster -  is it something like warmed up/cached version? Anyway, I picked up second round results in both cases (prevent and today's version)

Reply:
It is due to the `--fork 2` and is to show that there can be fluctuations between runs.